### PR TITLE
CT-2886: allowIncrement flag introduced

### PR DESCRIPTION
**Pull request by RafalOzogJudopay**

### CODE REVIEW

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleApp/Settings.bundle/Root.plist [hunk 1]
```diff
@@ -226,14 +226,14 @@
...
 		</dict>
-        <dict>
-            <key>Type</key>
-            <string>PSTextFieldSpecifier</string>
-            <key>Title</key>
-            <string>recommendation_api_timeout_title</string>
-            <key>Key</key>
-            <string>recommendation_api_timeout</string>
-        </dict>
+		<dict>
```
Comment (RafalOzogJudopay) [inline Examples/ObjectiveCExampleApp/ObjectiveCExampleApp/Settings.bundle/Root.plist:229]: This is clang change that actually fixed margin for this one dict element (not related to this PR, but good anyway).

File: Source/Models/Request/JPPreAuthApplePayRequest.m [hunk 2]
```diff
@@ -46,6 +48,7 @@ - (nonnull instancetype)initWithConfiguration:(nonnull JPConfiguration *)configu
                                    andPayment:(nonnull PKPayment *)payment {
     if (self = [super initWithConfiguration:configuration andPayment:payment]) {
         _delayedAuthorisation = configuration.isDelayedAuthorisation;
+        _allowIncrement = configuration.isAllowIncrementOn;
     }
```
Comment (RafalOzogJudopay) [inline Source/Models/Request/JPPreAuthApplePayRequest.m:52]: As headers of functions didn't change, I believe this solution is okay.

**Comment by RafalOzogJudopay** [comment]
The one thing we were about to discuss (you wanted to have a look first @stefan-tudor) is how to assign default value for the new boolean. Let's not forget about it before merging.

File: Source/Models/Configuration/JPConfiguration.h [hunk 3]
```diff
@@ -124,6 +124,8 @@
 
 @property (nonatomic, assign) BOOL isDelayedAuthorisation;
 
+@property (nonatomic, assign) BOOL isAllowIncrementOn;
```
Comment (stefan-tudor) [inline Source/Models/Configuration/JPConfiguration.h:127]: For consistency, let's rename this to `isAllowIncrement`, there's no other flag suffixed with `On`

File: Source/Services/ConfigurationValidation/JPConfigurationValidationService.m [hunk 4]
```diff
@@ -55,6 +55,7 @@ - (JPError *)validateConfiguration:(JPConfiguration *)configuration
     [self checkAmount:configuration.amount transactionType:transactionType error:&error];
     [self checkForValidJudoId:configuration error:&error];
     [self checkIfConsumerReferenceIsValid:configuration error:&error];
+    [self checkPreAuthPropertiesConfiguration:configuration transactionType:transactionType error:&error];
```
Comment (stefan-tudor) [inline Source/Services/ConfigurationValidation/JPConfigurationValidationService.m:58]: simply `checkPreAuthConfiguration`
and passing `transactionType` is redundant, simply add the check 
```
if transactionType == JPTransactionTypePreAuth then
   [checkPreAuthConfiguration:error:]
```

File: Source/Services/ConfigurationValidation/JPConfigurationValidationService.m [hunk 5]
```diff
@@ -221,6 +222,15 @@ - (void)checkAmount:(JPAmount *)amount transactionType:(JPTransactionType)transa
     }
 }
 
+- (void)checkPreAuthPropertiesConfiguration:(JPConfiguration *)configuration transactionType:(JPTransactionType)transactionType error:(NSError **)error {
+    BOOL isSupportedTransactionType = transactionType == JPTransactionTypePreAuth;
+    BOOL arePreAuthPropertiesInConflict = configuration.isDelayedAuthorisation == true && configuration.isAllowIncrementOn == true;
+
+    if (isSupportedTransactionType && arePreAuthPropertiesInConflict) {
```
Comment (stefan-tudor) [inline Source/Services/ConfigurationValidation/JPConfigurationValidationService.m:229]: These can be collapsed to:
```
if (configuration.isDelayedAuthorisation && configuration.isAllowIncrement)
```

**Review by stefan-tudor — CHANGES_REQUESTED** [review]
It looks good overall; excellent job!
Just a few minor remarks

File: Source/Models/Request/JPPreAuthApplePayRequest.m (hunk 2 above)
Comment (stefan-tudor) [reply Source/Models/Request/JPPreAuthApplePayRequest.m:52]: yup, looks good

**Comment by stefan-tudor** [comment]
> The one thing we were about to discuss (you wanted to have a look first @stefan-tudor) is how to assign default value for the new boolean. Let's not forget about it before merging.

isn't it already sending pre auth requests with this property as false?

File: Source/Models/Configuration/JPConfiguration.h (hunk 3 above)
Comment (RafalOzogJudopay) [reply Source/Models/Configuration/JPConfiguration.h]: Adjusted

File: Source/Services/ConfigurationValidation/JPConfigurationValidationService.m (hunk 4 above)
Comment (RafalOzogJudopay) [reply Source/Services/ConfigurationValidation/JPConfigurationValidationService.m]: Adjusted

File: Source/Services/ConfigurationValidation/JPConfigurationValidationService.m (hunk 5 above)
Comment (RafalOzogJudopay) [reply Source/Services/ConfigurationValidation/JPConfigurationValidationService.m]: Adjusted

**Comment by RafalOzogJudopay** [comment]
> > The one thing we were about to discuss (you wanted to have a look first @stefan-tudor) is how to assign default value for the new boolean. Let's not forget about it before merging.
> 
> isn't it already sending pre auth requests with this property as false?

Yes, it seems you're right. I've just tested this with debugger (commented out line that sets this boolean in configuration, and further in the code it had "NO" as a value).

**Review by stefan-tudor — APPROVED** [review]

**Pull request merged**